File: 20180503/lyrichu_20180503_01.java
/*
*date    : 2018-05-05 14:33:08
*author  : lyrichu
*email   :919987476@qq.com
*link    : https://www.github.com/Lyrichu
*version : 0.1
*description:
*Q1:
求数组的最长公共递增子序列
输入:一个无序数组
输出:最长公共递增子序列长度以及递增子序列
*/
import java.util.Stack;

public class lyrichu_20180503_01
{
	public static int lcs(int [] arr)
	{
		int [] arr1 = new int[arr.length]; //arr1 存放各元素作为最大元素的最长递增子序列的长度
		// pre　记录以该元素作为最大元素的递增序列中，该元素的前驱元素
		int [] pre = new int[arr.length];
		for(int i=0;i<arr.length;++i)
		{
			arr1[i] = 1;
			pre[i] = i;
		}
		for(int j=1;j<arr.length;++j)
		{
			for(int i=0;i<j;i++)
			{
				if(arr[j] > arr[i] && arr1[j] < arr1[i]+1)
				{
					arr1[j] = arr1[i] + 1;
					pre[j] = i;
				}
			}
		}
		int max = 0,index = 0;
		for(int k=0;k<arr1.length;++k)
		{
			if(arr1[k]>max)
			{
				max = arr1[k];
				index = k;
			}
		}
		System.out.println("最长递增子序列长度为:" + max);
		Stack<Integer> stack = new Stack<Integer>();
		while(index != pre[index])
		{
			stack.push(arr[index]);
			index = pre[index];
		}
		if(index == pre[index])
		{
			stack.push(arr[index]);
		}
		while (!stack.isEmpty()){
			System.out.print(stack.pop()+" ");
		}
		System.out.println();
		return max;
	}
	public static void main(String args[])
	{
		int [] arr = {35, 36, 39, 3, 15, 27, 6, 42};
		lcs(arr);
	}
}

File: 20180503/lyrichu_20180503_01.c
/*
*date    : 2018-05-05 13:43:51
*author  : lyrichu
*email   :919987476@qq.com
*link    : https://www.github.com/Lyrichu
*version : 0.1
*description:
*求数组的最长公共递增子序列
输入:一个无序数组
输出:最长公共递增子序列长度
思路:将问题转化为原数组和其递增数组的最长公共子序列的问题
*/
#include <stdio.h>
#include <stdlib.h>
# define MAXN 100 //数组最大长度
int arr1[MAXN]; //原数组
int arr2[MAXN]; // 排序数组
int N; //数组长度
int dp[MAXN+1][MAXN+1]; // dp[i][j]表示arr1前i子数组和arr2前j子数组的最长公共子序列长度

int cmp(const void *a,const void *b)
{
	return *(int*)a - *(int*)b;  // 递增排序
}

void lcs()
{
	for(int i=0;i<=N;i++)
	{
		for(int j=0;j<=N;j++)
		{
			if(i==0 || j==0)
				dp[i][j] = 0;
			else if(arr1[i-1] == arr2[j-1])
				dp[i][j] = dp[i-1][j-1] + 1;
			else
				dp[i][j] = (dp[i-1][j] > dp[i][j-1] ? dp[i-1][j]:dp[i][j-1]);
		}
	}
}

int main(int argc, char const *argv[])
{
	scanf("%d",&N);
	for(int i=0;i<N;i++)
	{
		scanf("%d",&arr1[i]);
		arr2[i] = arr1[i];
	}
	// 对arr2进行排序
	qsort(arr2,N,sizeof(arr2[0]),cmp);
	//初始化dp为0
	for(int i=0;i<=N;++i)
		for(int j=0;j<=N;++j)
			dp[i][j] = 0;
	lcs();
	printf("%d\n",dp[N][N]);
	return 0;
}

